src/hpack/decoder.rs

```diff
@@ -292,7 +295,11 @@ impl Decoder {
         }
     }
 
-    fn decode_string(&mut self, buf: &mut Cursor<&mut BytesMut>) -> Result<Bytes, DecoderError> {
+    fn try_decode_string(
+        &mut self,
+        buf: &mut Cursor<&mut BytesMut>,
+    ) -> Result<((usize, usize), Option<Bytes>), DecoderError> {
```

Comment:
It's hard to know what these values are supposed to be. What if instead of returning a tuple, it return some `struct DecodedString { .. }` that named the `usize`s and `Option<Bytes>`?

Reply:
Thank you for reply.

I am using tonic as grpc client and would receive a large HEADERS frame(`grpc-status-details-bin`) whose name is non_huff format and value is partial from C++ server. When decoding this header, the name is consumed because it is not huff, and then the value decoding is failed for `DecoderError::NeedMore`, and then the decode function returns with error but the buf is consumed.

I see that the CONTINUATION header currently tested in CI but both name and value are huff_encoded.

My first patch is simply replacing `consume()` with buf copy, and @[olix0r](https://github.com/olix0r) pointed that the copy can be optimized. So i tried to save the decoding context and consume the buf only if both header name and header buf is successfully decoded.

I will update my patch to easily show the meaning of those values.